### PR TITLE
[JENKINS-38992] Allow caching library versions

### PR DESCRIPTION
**Pull request by julienduchesne**
We have a jenkins instance that need to pull a lot of libraries. When many jobs are running simultaneously it can sometimes take up to 30 seconds to load a single library.

Probably fixes https://issues.jenkins-ci.org/browse/JENKINS-38992 and https://issues.jenkins-ci.org/browse/JENKINS-48210 in some measure

### CODE REVIEW

**Comment by julienduchesne** [comment]
Alright, and I still have to fix failing tests. FYI, that's how it looks in the UI 
image

**Comment by julienduchesne** [comment]
Added tests, the build failure seems unrelated:
```java.lang.RuntimeException: Error grabbing Grapes -- [download failed: org.apache.commons#commons-math3;3.4.1!commons-math3.jar]```

**Comment by julienduchesne** [comment]
Any update on this?

**Comment by julienduchesne** [comment]
Hey @dwnusbaum. I've fixed every issue except putting the config in SCMSourceRetriever. I put the config in LibraryConfiguration on purpose, that way it is independant from the SCM. The same way that other configs such as the default version and default version override are. I also don't think that it's a bug if I change the repository but the version stays the same. Let me know what you think.

File: src/main/java/org/jenkinsci/plugins/workflow/libs/LibraryConfiguration.java
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/libs/LibraryConfiguration.java]: Please avoid touching lines of text with only whitespace changes, so as to keep diffs minimal. You can configure your text editor to refrain from doing so.

File: src/main/resources/org/jenkinsci/plugins/workflow/libs/LibraryCachingConfiguration/help-excludedVersionsStr.html
Comment (jglick) [inline src/main/resources/org/jenkinsci/plugins/workflow/libs/LibraryCachingConfiguration/help-excludedVersionsStr.html]: But there is no possibility here to specify a regular expression.

FWIW my proposal in JENKINS-38992 was to make this automatic—zero configuration. The SCM should know whether or not a given version was deterministic. The current `SCMSource.retrieve(String, TaskListener)` does not suffice since at least in the case of `AbstractGitSCMSource` the resulting `SCMRevision.isDeterministic` will always be true—the method resolves the actual current head of a branch. Thus we would need to extend the `SCMHead` API in `scm-api` to let the SCM source indicate whether the _head_ is deterministic (like a hash, or arguably a tag) vs. nondeterministic (like a branch). This would be easy to implement in `git` at least.

Comment (julienduchesne) [reply src/main/resources/org/jenkinsci/plugins/workflow/libs/LibraryCachingConfiguration/help-excludedVersionsStr.html]: I agree that it should support regex and maybe add an include field as well (only include a single version) but is it really user friendly to cache some versions but not others. In our use case, it would be annoying to cache tags by default because we use semantic versioning and publish minor versions as well as patch versions (ex: release v4.2.1 also releases v4.2 and v4).

All I'm saying is I'm not a fan of hidden behavior and the current implementation is very simple. Is it really necessary to complicate it?

**Comment by jennbriden** [comment]
I have had several customers report hitting performance issues due to shared libraries not being cached. What LTS do we think this will go into?

**Comment by julienduchesne** [comment]
It works very well though. We are using it on a jenkins instance running about 20 very short jobs (15 sec to 5 minutes) concurrently at all times during the day. We used to only be able to do a full cycle of our workflow in over an hour before this. It now takes under 20 minutes.

**Comment by criley-eagle** [comment]
> "Based on input here, this might not be the best implementation."
@jennbriden Is this pr still viable, or is there a better implementation in the works, as we are definitely hitting up against the same issue this pr was attempting to resolve?

**Comment by julienduchesne** [comment]
I'm keeping this updated in case you guys ever want to merge it :). It's been working fine without a single glitch since I started using it and our job load has about doubled since then.

**Comment by julienduchesne** [comment]
Would you guys be willing to take another look at this? It might not (arguably) be the absolute best solution but it works flawlessly and it is the most simple. How can we get this merged?

**Comment by ppattard** [comment]
+1 
Would love to have this story delivered. We are heavily relying on pipeline libraries and fetching them at each build is costly + introduces a point of failure on the repository we fetch from.

**Comment by ajrussellsap** [comment]
I think this is a really important improvement that would improve build times for a lot of people.  Even if the implementation might not be 100% perfect, the benefits outweigh any downsides.  Please can we get this merged?

**Comment by kmurrayAtCloudbees** [comment]
Is there a plan to merge this anytime soon?

**Comment by rpannek** [comment]
We would need some solution for caching libraries too because we're moving our from our internal infra to github because we're open sourcing it.

**Comment by julienduchesne** [comment]
This PR is battle-tested by now. We have run over 200k jobs with it (No exaggeration, I archived most of build history not too long ago). Honestly, without this, Jenkins would just be unusable for us. What changes would be needed for this to be merged?

**Comment by Inetov** [comment]
@bitwiseman @dwnusbaum guys, can you help with the review?

**Comment by MarkEWaite** [comment]
@oleg-nenashev the [CI job is failing on ci.jenkins.io](https://ci.jenkins.io/job/Plugins/job/workflow-cps-global-lib-plugin/job/PR-50/) because the JGit version provided in git plugin 3.12.1 cannot merge the changes from this branch with the current master branch.  

The version of JGit included in the latest release of the git plugin (4.1.1) and git client plugin (3.1.1) **can merge** the two branches.  Command line git **can merge** the two branches as well.  The merge failure is specific to JGit 4.7.0.

The original submitters could resolve the issue by merging the master branch to their branch.  We could resolve the issue by updating the plugins on ci.jenkins.io.  I'll check with @olblak on the policy for plugin updates on ci.jenkins.io.

**Comment by patrickhuy** [comment]
@bitwiseman @dwnusbaum @oleg-nenashev it seems like this would give great improvements to a lot of Jenkins users. Is there any chance on moving this forward?

**Comment by oleg-nenashev** [comment]
I pinged the maintainers multiple times over past months. Unfortunately I cannot  do much more about it, but I will try pinging again

**Comment by julienduchesne** [comment]
:wave: I haven't received any "pings" until now. What needs to be done to have this merged?

**Comment by msrb** [comment]
@oleg-nenashev LGTM? :)

**Comment by oleg-nenashev** [comment]
> 👋 I haven't received any "pings" until now. What needs to be done to have this merged?

Looks like somebody needs to take ownership of the plugin or to get the current maintainers to review/merge the pull request. Unfortunately I cannot do much more to push it forward at the moment.

**Comment by atmcarmo** [comment]
Thanks for this PR! Hope this gets merged. It's really an issue for us the time it gets to clone and load a library when many pipelines are starting to execute at the same time.

**Comment by robodenitro** [comment]
any updates here?

**Comment by MarkEWaite** [comment]
@robodenitro I've been running this pull request since early November 2020 without detecting any issue.  You could do the same by installing the file from https://github.com/MarkEWaite/docker-lfs/commit/408684e046549fb3b8539ec5ded85006fdde11ab .  Once you do that, your specific need should be resolved and you can report your test results

It is more persuasive to report successful use of a pull request than to ask for a status update.

**Comment by nnadams** [comment]
Following Mark's example, I have been using this pull request since June 2020 and seen no issues. We run thousands of jobs weekly that all make use of this caching. Our CI has been faster and much more robust. I hope this can be merged. Either way thank you @julienduchesne!

File: src/test/java/org/jenkinsci/plugins/workflow/libs/ResourceStepTest.java
Comment (MarkEWaite) [inline src/test/java/org/jenkinsci/plugins/workflow/libs/ResourceStepTest.java:61]: Remove trailing white space
```suggestion

```

File: src/test/java/org/jenkinsci/plugins/workflow/libs/ResourceStepTest.java
Comment (MarkEWaite) [inline src/test/java/org/jenkinsci/plugins/workflow/libs/ResourceStepTest.java:93]: Remove trailing white space
```suggestion

```

File: src/test/java/org/jenkinsci/plugins/workflow/libs/ResourceStepTest.java
Comment (MarkEWaite) [inline src/test/java/org/jenkinsci/plugins/workflow/libs/ResourceStepTest.java:72]: Remove trailing white space
```suggestion

```

File: src/test/java/org/jenkinsci/plugins/workflow/libs/ResourceStepTest.java
Comment (MarkEWaite) [inline src/test/java/org/jenkinsci/plugins/workflow/libs/ResourceStepTest.java:115]: Remove trailing white space
```suggestion

```

**Review by MarkEWaite — APPROVED** [review]
Remove trailing white space in diffs

**Comment by ababushk** [comment]
We use version of workflow-cps-global-lib with changes from this PR on production instance with huge amount of jobs and builds per day. No issues so far!
Moreover, this allowed us to run multiple concurrent builds (>1000 simultaneously) without waiting for a lock on git repo cache directory for our global library on Jenkins server, which has speeded up them drastically. Many thanks to @julienduchesne !

**Comment by atmcarmo** [comment]
This solution seems to be rock solid, really improving performance even with a very big amount of builds happening at the same time.

What needs to be done to merge to be included in a new release?

Thank you very much for your work 👏

**Comment by michelgasser** [comment]
Why is the PR not being merged? Why is it taking so long?

**Comment by timja** [comment]
@dwnusbaum @car-roll @bitwiseman Would it be possible for someone to take a look at this?

**Comment by oleg-nenashev** [comment]
As a CloudBees employee I spent so much time pinging the current
maintainers. Doing the last attempt

**Review by car-roll — APPROVED** [review]
So, I'm just going over the comments here and I think this summarizes what went on here:

It does look like most issues were addressed, but there was a preference by both @dwnusbaum and @jglick to implement things slightly different, which is why this PR stalled. But because there was no significant effort to address the alternatives proposed by them, and since this PR was the ONLY solution available, it became widely adopted. Thus the wide adoption of this solution is now causing pressure for this PR to be merged. 

Given that this PR has been actively used in the wild now for quite some time, I'm going ahead to merge it.

**Comment by michelgasser** [comment]
Cool, thanks a lot for the merge! :-)

**Comment by MaximeAnsquer** [comment]
How can I disable caching please?

**Comment by MaximeAnsquer** [comment]
Thank you @MarkEWaite!

**Comment by aaronr93** [comment]
Hi @MarkEWaite, just following up on my first question above. Is there a way to programmatically clear the cache from a pipeline?

**Comment by MarkEWaite** [comment]
> Hi @MarkEWaite, just following up on my first question above. Is there a way to programmatically clear the cache from a pipeline?

No method that I know to clear cache from a Pipeline.

**Pull request merged**
